src/ControllerMessenger.ts

```diff
@@ -0,0 +1,89 @@
+export type ActionHandler<Action, ActionType> = (
+  ...args: ExtractActionParameters<Action, ActionType>
+) => ExtractActionResponse<Action, ActionType>;
+export type ExtractActionParameters<Action, T> = Action extends { type: T; handler: (...args: infer H) => any }
+  ? H
+  : never;
+export type ExtractActionResponse<Action, T> = Action extends { type: T; handler: (...args: any) => infer H }
+  ? H
+  : never;
+
+export type ExtractEvenHandler<Event, T> = Event extends { type: T; payload: infer P }
+  ? P extends any[]
+    ? (...payload: P) => void
+    : never
+  : never;
+export type ExtractEventPayload<Event, T> = Event extends { type: T; payload: infer P } ? P : never;
+
+type ActionConstraint = { type: string; handler: (...args: any) => unknown };
```

Comment:
I wasn't able to get rid of this `any` :thinking: Using `unknown` or `unknown[]` seemed to upset TypeScript. What I wanted here was to express that I don't know anything about these parameters, but I don't know how to do that.

Though, maybe it's fine to leave this as-is. There are unit tests for zero parameters, one parameter, and multiple parameters (of various different types). Those should be sufficient for ensuring we're not assuming anything about these parameters in this class.

Reply:
When i was trying to figure out types I had to do some 'any' to allow it to work, even though I had full type safety in practice. Hard to explain/understand why that was true but I think it's fine as is